File: sdk/java/src/test/java/jp/kengos/shojiku/ExternalSignerTest.java
package jp.kengos.shojiku;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.KeyFactory;
import java.security.PrivateKey;
import java.security.Signature;
import java.security.spec.PKCS8EncodedKeySpec;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.UnaryOperator;
import org.junit.jupiter.api.Test;

/**
 * Signing with a key this process is never given.
 *
 * <p>The engine hands out bytes, something else signs them, and the finished document has to
 * verify. Nothing is stubbed: the function here signs with the JDK's own crypto over a key this
 * package never hands to the engine, which is exactly the shape a cloud key service takes from its
 * point of view.
 */
class ExternalSignerTest extends Fixtures {

  @Test
  void signingWithAKeyHeldElsewhereProducesADocumentThatVerifies() throws Exception {
    ShojikuClient client = client().build();
    DocumentArtifact rendered = rendered(client);

    Result<DocumentArtifact> result = client.sign(rendered, external("rsa2048"));

    assertTrue(result.success(), () -> String.valueOf(result.failure()));
    // Append-only: the signed bytes begin with the input byte for byte.
    byte[] signed = result.artifact().bytes();
    assertArrayEquals(rendered.bytes(), java.util.Arrays.copyOf(signed, rendered.bytes().length));
    assertTrue(client.verify(result.artifact(), List.of(key("rsa2048.cert.pem")), null).success());
  }

  @Test
  void signingWithAnEllipticCurveKey() throws Exception {
    ShojikuClient client = client().build();

    Result<DocumentArtifact> result =
        client.sign(rendered(client), external("ec256", Algorithm.ECDSA_P256_SHA256));

    assertTrue(result.success(), () -> String.valueOf(result.failure()));
  }

  @Test
  void theFunctionIsHandedTheSignedAttributesNotTheDocumentDigest() throws Exception {
    // The distinction the shorthand gets wrong: signing the digest instead
    // produces a document that fails verification.
    List<byte[]> seen = new ArrayList<>();
    UnaryOperator<byte[]> inner = signing("rsa2048", Algorithm.RSA_PKCS1_SHA256);
    ShojikuClient client = client().build();

    client.sign(
        rendered(client),
        ExternalSigner.of(
            toBeSigned -> {
              seen.add(toBeSigned);
              return inner.apply(toBeSigned);
            },
            Algorithm.RSA_PKCS1_SHA256,
            key("rsa2048.cert.pem")));

    assertEquals(1, seen.size());
    // A DER SET OF attributes (RFC 5652's explicit form, tag 0x31), not the
    // 32-byte SHA-256 digest.
    assertEquals((byte) 0x31, seen.get(0)[0]);
    assertNotEquals(32, seen.get(0).length);
  }

  @Test
  void aCertificateHeldInMemoryNeverHasToBeWrittenDown() throws Exception {
    ShojikuClient client = client().build();
    ExternalSigner provider =
        ExternalSigner.ofPem(
            signing("rsa2048", Algorithm.RSA_PKCS1_SHA256),
            Algorithm.RSA_PKCS1_SHA256,
            Files.readAllBytes(key("rsa2048.cert.pem")));

    assertTrue(client.sign(rendered(client), provider).success());
    assertTrue(provider.toString().contains("[pem bytes]"));
  }

  @Test
  void aSignatureWithNothingInItIsRefused() {
    ShojikuClient client = client().build();
    ExternalSigner provider =
        ExternalSigner.of(
            toBeSigned -> new byte[0], Algorithm.RSA_PKCS1_SHA256, key("rsa2048.cert.pem"));

    UsageException error =
        assertThrows(UsageException.class, () -> client.sign(rendered(client), provider));
    assertTrue(error.getMessage().contains("non-empty signature"));
  }

  @Test
  void aFunctionThatReturnsNothingAtAllIsRefusedToo() {
    // Null and empty are the same claim — there is no signature — and both
    // have to be refused rather than written into a document.
    ShojikuClient client = client().build();
    ExternalSigner provider =
        ExternalSigner.of(toBeSigned -> null, Algorithm.RSA_PKCS1_SHA256, key("rsa2048.cert.pem"));

    UsageException error =
        assertThrows(UsageException.class, () -> client.sign(rendered(client), provider));
    assertTrue(error.getMessage().contains("non-empty signature"));
  }

  @Test
  void theFunctionsOwnFailureIsNotFiledAsADocumentFailure() {
    // A key service outage is the caller's, not a fact about this document.
    ShojikuClient client = client().build();
    ExternalSigner provider =
        ExternalSigner.of(
            toBeSigned -> {
              throw new IllegalStateException("the key service is unreachable");
            },
            Algorithm.RSA_PKCS1_SHA256,
            key("rsa2048.cert.pem"));

    IllegalStateException error =
        assertThrows(IllegalStateException.class, () -> client.sign(rendered(client), provider));
    assertEquals("the key service is unreachable", error.getMessage());
  }

  @Test
  void noSignatureIsAskedForWhenPreparingFailed() {
    // An unreadable certificate is a fact about the inputs; paying for a
    // signature afterwards would tell the caller nothing new.
    AtomicBoolean asked = new AtomicBoolean();
    ShojikuClient client = client().build();
    ExternalSigner provider =
        ExternalSigner.of(
            toBeSigned -> {
              asked.set(true);
              return new byte[] {1};
            },
            Algorithm.RSA_PKCS1_SHA256,
            Path.of("/nonexistent/signer.crt"));

    Result<DocumentArtifact> result = client.sign(rendered(client), provider);

    assertTrue(result.failed());
    assertFalse(asked.get());
  }

  @Test
  void aRefusedDocumentComesBackAsAFailedResultWithNoSignatureAsked() {
    // The engine itself refuses: these bytes are not a document it rendered.
    AtomicBoolean asked = new AtomicBoolean();
    ShojikuClient client = client().build();
    ExternalSigner provider =
        ExternalSigner.of(
            toBeSigned -> {
              asked.set(true);
              return new byte[] {1};
            },
            Algorithm.RSA_PKCS1_SHA256,
            key("rsa2048.cert.pem"));

    Result<DocumentArtifact> result =
        client.sign(client.artifact("not a PDF".getBytes(StandardCharsets.UTF_8)), provider);

    assertTrue(result.failed());
    assertFalse(asked.get());
  }

  @Test
  void theCertificateAndTheFunctionAreTakenExplicitly() {
    UnaryOperator<byte[]> sign = bytes -> new byte[] {1};

    assertTrue(
        assertThrows(
                UsageException.class,
                () -> ExternalSigner.of(null, Algorithm.RSA_PKCS1_SHA256, key("rsa2048.cert.pem")))
            .getMessage()
            .contains("function that signs"));
    assertTrue(
        assertThrows(
                UsageException.class, () -> ExternalSigner.of(sign, null, key("rsa2048.cert.pem")))
            .getMessage()
            .contains("needs an algorithm"));
    assertTrue(
        assertThrows(
                UsageException.class,
                () -> ExternalSigner.of(sign, Algorithm.RSA_PKCS1_SHA256, null))
            .getMessage()
            .contains("needs either"));
    assertTrue(
        assertThrows(
                UsageException.class,
                () -> ExternalSigner.ofPem(sign, Algorithm.RSA_PKCS1_SHA256, null))
            .getMessage()
            .contains("needs either"));
  }

  @Test
  void aPayloadThatNamesNoBytesToSignIsRefused() {
    // The real engine always reports them, so this is the shape only a
    // different library on the other end could produce.
    assertTrue(
        assertThrows(UsageException.class, () -> ExternalSigner.bytesToSign("{}"))
            .getMessage()
            .contains("no bytes to sign"));
    assertTrue(
        assertThrows(UsageException.class, () -> ExternalSigner.bytesToSign("{\"toBeSigned\":7}"))
            .getMessage()
            .contains("no bytes to sign"));
    assertArrayEquals(
        "123".getBytes(StandardCharsets.UTF_8),
        ExternalSigner.bytesToSign("{\"toBeSigned\":\"MTIz\"}"));
  }

  @Test
  void thePrintedFormShowsTheCertificateFormAndTheAlgorithmOnly() throws Exception {
    String shown = external("ec256", Algorithm.ECDSA_P256_SHA256).toString();

    assertTrue(shown.contains("ec256.cert.pem"));
    assertTrue(shown.contains("ecdsa-p256-sha256"));
    assertEquals(
        Algorithm.ECDSA_P256_SHA256, external("ec256", Algorithm.ECDSA_P256_SHA256).algorithm());
    assertFalse(shown.contains("Lambda"));
  }

  @Test
  void aRegisteredExternalSignerSignsFromAStrictClient() throws Exception {
    // The provider a strict deployment may use is a NAMED one, and an external
    // signer is as nameable as a local key.
    ShojikuClient client = client().providers(Map.of("kms", external("rsa2048"))).build();

    assertTrue(client.sign(rendered(client), "kms").success());
  }

  @Test
  void aBareExternalSignerIsRefusedByAStrictClient() throws Exception {
    ShojikuClient strict =
        client().strict(true).providers(Map.of("kms", external("rsa2048"))).build();

    UsageException error =
        assertThrows(
            UsageException.class,
            () -> strict.sign(rendered(client().build()), external("rsa2048")));
    assertTrue(error.getMessage().contains("registered in configuration"));
  }

  @Test
  void aRegisteredValueThatCarriesNoMaterialIsRefused() {
    // The registry holds SigningProvider, a public marker: a value implementing
    // it and nothing else would otherwise reach the transport with no way to
    // produce a signature.
    SigningProvider marker = new SigningProvider() {};
    ShojikuClient client = client().providers(Map.of("kms", marker)).build();

    assertTrue(
        assertThrows(UsageException.class, () -> client.sign(rendered(client), "kms"))
            .getMessage()
            .contains("no signing provider named"));
  }

  private ExternalSigner external(String stem) throws IOException {
    return external(stem, Algorithm.RSA_PKCS1_SHA256);
  }

  private ExternalSigner external(String stem, Algorithm algorithm) throws IOException {
    return ExternalSigner.of(signing(stem, algorithm), algorithm, key(stem + ".cert.pem"));
  }

  /**
   * A stand-in for a key service: signs with a key this package never hands to the engine. The
   * output is the raw operation's, which is what both major cloud key services return.
   */
  private UnaryOperator<byte[]> signing(String stem, Algorithm algorithm) throws IOException {
    byte[] der = pkcs8(Files.readString(key(stem + ".key.pem")));
    String keyAlgorithm = algorithm == Algorithm.RSA_PKCS1_SHA256 ? "RSA" : "EC";
    String signatureAlgorithm =
        algorithm == Algorithm.RSA_PKCS1_SHA256 ? "SHA256withRSA" : "SHA256withECDSA";
    return toBeSigned -> {
      try {
        PrivateKey privateKey =
            KeyFactory.getInstance(keyAlgorithm).generatePrivate(new PKCS8EncodedKeySpec(der));
        Signature signature = Signature.getInstance(signatureAlgorithm);
        signature.initSign(privateKey);
        signature.update(toBeSigned);
        return signature.sign();
      } catch (Exception error) {
        throw new IllegalStateException("the stand-in key service failed", error);
      }
    };
  }

  private static byte[] pkcs8(String pem) {
    String body =
        pem.replace("-----BEGIN PRIVATE KEY-----", "")
            .replace("-----END PRIVATE KEY-----", "")
            .replaceAll("\\s", "");
    return Base64.getDecoder().decode(body);
  }
}

File: sdk/js/test/externalSigner.test.ts
/**
 * Signing with a key this process is never given.
 *
 * The engine hands out bytes, something else signs them, and the finished
 * document has to verify. Nothing is stubbed: the callback here runs `openssl`
 * over the bytes it is handed, which is exactly the shape a cloud key service
 * takes from this package's point of view.
 */

import { execFile } from 'node:child_process';
import { mkdtemp, readFile, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { inspect, promisify } from 'node:util';
import { afterEach, describe, expect, it } from 'vitest';
import { bytesToSign } from '../src/externalSigner.js';
import type { Algorithm } from '../src/index.js';
import { ExternalSigner, resetConfiguration, UsageError } from '../src/index.js';
import { keyPath, makeClient, rendered } from './support/fixtures.js';

afterEach(resetConfiguration);

const run = promisify(execFile);

/**
 * A stand-in for a key service: signs with a key this package never sees.
 *
 * `openssl dgst -sha256 -sign` produces exactly what the engine expects —
 * PKCS#1 v1.5 bytes for an RSA key, an ASN.1 DER sequence for an EC one —
 * which is also what AWS KMS and Google Cloud KMS return.
 */
function opensslSigner(stem: string): (toBeSigned: Buffer) => Promise<Buffer> {
  return async (toBeSigned) => {
    const dir = await mkdtemp(join(tmpdir(), 'shojiku-js-sign-'));
    try {
      const message = join(dir, 'to-be-signed.bin');
      const signature = join(dir, 'signature.bin');
      await writeFile(message, toBeSigned);
      await run('openssl', [
        'dgst',
        '-sha256',
        '-sign',
        keyPath(`${stem}.key.pem`),
        '-out',
        signature,
        message,
      ]);
      return await readFile(signature);
    } finally {
      await rm(dir, { recursive: true, force: true });
    }
  };
}

function external(stem = 'rsa2048', algorithm: Algorithm = 'rsa-pkcs1-sha256'): ExternalSigner {
  return new ExternalSigner({
    cert: keyPath(`${stem}.cert.pem`),
    algorithm,
    sign: opensslSigner(stem),
  });
}

describe('signing with a key held elsewhere', () => {
  it('produces a document that verifies', async () => {
    const artifact = await rendered();

    const result = await artifact.sign(external());

    expect(result.success).toBe(true);
    const signed = result.unwrap();
    // Append-only: the input is a byte-for-byte prefix of the output.
    expect(signed.bytes.subarray(0, artifact.bytes.length)).toEqual(artifact.bytes);
    const verified = await signed.verify({ anchors: keyPath('rsa2048.cert.pem') });
    expect(verified.success).toBe(true);
  });

  it('signs with an elliptic-curve key as well', async () => {
    const result = await (await rendered()).sign(external('ec256', 'ecdsa-p256-sha256'));

    expect(result.success).toBe(true);
  });

  it('hands the callback the signed ATTRIBUTES, not the document digest', async () => {
    // The distinction the shorthand gets wrong: signing the digest instead
    // produces a document that fails verification.
    const seen: Buffer[] = [];
    const inner = opensslSigner('rsa2048');
    const provider = new ExternalSigner({
      cert: keyPath('rsa2048.cert.pem'),
      algorithm: 'rsa-pkcs1-sha256',
      sign: async (toBeSigned) => {
        seen.push(toBeSigned);
        return inner(toBeSigned);
      },
    });

    await (await rendered()).sign(provider);

    expect(seen).toHaveLength(1);
    // A DER SET OF attributes (RFC 5652's explicit form, tag 0x31), not the
    // 32-byte SHA-256 digest.
    expect(seen[0]?.[0]).toBe(0x31);
    expect(seen[0]?.length).not.toBe(32);
  });

  it('takes a certificate held in memory, so it never has to be written down', async () => {
    const provider = new ExternalSigner({
      certPem: await readFile(keyPath('rsa2048.cert.pem')),
      algorithm: 'rsa-pkcs1-sha256',
      sign: opensslSigner('rsa2048'),
    });

    expect((await (await rendered()).sign(provider)).success).toBe(true);
    expect(inspect(provider)).toContain('[pem bytes]');
  });
});

describe('what counts as misuse', () => {
  it('refuses a callback that returns nothing to write', async () => {
    const provider = new ExternalSigner({
      cert: keyPath('rsa2048.cert.pem'),
      algorithm: 'rsa-pkcs1-sha256',
      sign: () => Buffer.alloc(0),
    });

    await expect((await rendered()).sign(provider)).rejects.toThrow(UsageError);
  });

  it('refuses a callback that returns something other than bytes', async () => {
    const provider = new ExternalSigner({
      cert: keyPath('rsa2048.cert.pem'),
      algorithm: 'rsa-pkcs1-sha256',
      sign: (() => 'a signature, allegedly') as never,
    });

    await expect((await rendered()).sign(provider)).rejects.toThrow(/non-empty Buffer/);
  });

  it('needs a `sign` callback at all', () => {
    expect(() => new ExternalSigner({ cert: 'signer.crt', algorithm: 'rsa-pkcs1-sha256' })).toThrow(
      /`sign` callback/,
    );
  });

  it('takes its certificate explicitly, in both directions', () => {
    const sign = () => Buffer.from('x');
    expect(
      () =>
        new ExternalSigner({
          cert: 'signer.crt',
          certPem: Buffer.from('-----BEGIN CERTIFICATE-----'),
          algorithm: 'rsa-pkcs1-sha256',
          sign,
        }),
    ).toThrow(/not both/);
    expect(() => new ExternalSigner({ algorithm: 'rsa-pkcs1-sha256', sign })).toThrow(
      /needs either/,
    );
  });

  it('needs an algorithm, and names the accepted ones without echoing the request', () => {
    const sign = () => Buffer.from('x');
    expect(() => new ExternalSigner({ cert: 'signer.crt', sign })).toThrow(/needs `algorithm`/);

    let message = '';
    try {
      new ExternalSigner({ cert: 'signer.crt', algorithm: 'rsa-pkcs1-sha1' as never, sign });
    } catch (error) {
      message = String((error as Error).message);
    }
    expect(message).toContain('rsa-pkcs1-sha256');
    expect(message).toContain('ecdsa-p256-sha256');
    expect(message).not.toContain('sha1`');
  });
});

describe('what is not this package’s problem', () => {
  it('lets the callback’s own rejection out rather than filing it as a document failure', async () => {
    // A key service outage is the caller's, not a fact about this document.
    const provider = new ExternalSigner({
      cert: keyPath('rsa2048.cert.pem'),
      algorithm: 'rsa-pkcs1-sha256',
      sign: () => Promise.reject(new Error('the key service is unreachable')),
    });

    await expect((await rendered()).sign(provider)).rejects.toThrow(
      'the key service is unreachable',
    );
  });

  it('returns a failed prepare — a refused document — without asking for a signature', async () => {
    // The engine itself refuses: these bytes are not a document it rendered.
    let asked = false;
    const client = makeClient();
    const notADocument = client.artifact(Buffer.from('not a PDF at all'));
    const provider = new ExternalSigner({
      cert: keyPath('rsa2048.cert.pem'),
      algorithm: 'rsa-pkcs1-sha256',
      sign: () => {
        asked = true;
        return Buffer.from('never reached');
      },
    });

    const result = await client.sign(notADocument, provider);

    expect(result.failed).toBe(true);
    expect(asked).toBe(false);
  });

  it('refuses a prepare payload that names no bytes to sign', () => {
    // The real engine always reports them, so this is the shape only a
    // different program on the other end could produce.
    expect(() => bytesToSign('{}')).toThrow(/no bytes to sign/);
    expect(bytesToSign('{"toBeSigned":"MTIz"}').toString()).toBe('123');
  });

  it('returns an unreadable certificate without ever asking for a signature', async () => {
    // An unreadable certificate is a fact about the inputs; paying for a
    // signature afterwards would tell the caller nothing new.
    let asked = false;
    const provider = new ExternalSigner({
      cert: '/nonexistent/signer.crt',
      algorithm: 'rsa-pkcs1-sha256',
      sign: () => {
        asked = true;
        return Buffer.from('never reached');
      },
    });

    const result = await (await rendered()).sign(provider);

    expect(result.failed).toBe(true);
    expect(asked).toBe(false);
  });
});

describe('what it prints', () => {
  it('shows the certificate form and the algorithm, and nothing else', () => {
    const provider = external('rsa2048', 'ecdsa-p256-sha256');

    const shown = inspect(provider);
    expect(shown).toContain('rsa2048.cert.pem');
    expect(shown).toContain('ecdsa-p256-sha256');
    expect(shown).not.toContain('sign');
    expect(JSON.stringify(provider)).not.toContain('function');
  });
});

describe('under a locked-down client', () => {
  it('signs when registered by name', async () => {
    // The provider a strict deployment may use is a NAMED one, and an
    // external signer is as nameable as a local key.
    const client = makeClient({ providers: { kms: external() } });

    expect((await client.sign(await rendered(), 'kms')).success).toBe(true);
  });

  it('is refused as a bare value when the client is strict', async () => {
    const client = makeClient({ strict: true, providers: { kms: external() } });

    await expect(client.sign(await rendered(), external())).rejects.toThrow(
      /registered in configuration/,
    );
  });
});
